### PR TITLE
Migrate client-go/metadata to contextual logging 

### PR DESCRIPTION
**Pull request by ricardoapl**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Migrate client-go/metadata to contextual logging

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/enhancements/issues/3077

#### Special notes for your reviewer:

Log output before migration

```
I1202 17:13:16.944886 1320000 metadata.go:194] Unable to retrieve PartialObjectMetadata: &runtime.notRegisteredErr{schemeName:"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme/register.go:27", gvk:schema.GroupVersionKind{Group:"", Version:"v1", Kind:"Pod"}, target:runtime.GroupVersioner(nil), t:reflect.Type(nil)}
I1202 17:25:19.610064 1472026 metadata.go:230] Unable to retrieve PartialObjectMetadataList: &runtime.notRegisteredErr{schemeName:"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme/register.go:27", gvk:schema.GroupVersionKind{Group:"", Version:"v1", Kind:"PodList"}, target:runtime.GroupVersioner(nil), t:reflect.Type(nil)}
```

Log output after migration

```
I1202 17:10:46.625898 1286529 metadata.go:194] "Could not retrieve PartialObjectMetadata" err="no kind \"Pod\" is registered for version \"v1\" in scheme \"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme/register.go:27\""
I1202 17:27:14.864645 1496151 metadata.go:230] "Could not retrieve PartialObjectMetadataList" err="no kind \"PodList\" is registered for version \"v1\" in scheme \"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme/register.go:27\""
```

#### Does this PR introduce a user-facing change?

```release-note
Migrate client-go/metadata to contextual logging
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/42bb993e369d166142b7181e90882066ad6c7651/keps/sig-instrumentation/3077-contextual-logging
```

/wg structured-logging 
/area logging
/priority important-longterm
/cc @kubernetes/wg-structured-logging-reviews

### CODE REVIEW

File: staging/src/k8s.io/client-go/metadata/metadata.go [hunk 1]
```diff
@@ -191,7 +191,7 @@ func (c *client) Get(ctx context.Context, name string, opts metav1.GetOptions, s
 	}
 	obj, err := result.Get()
 	if runtime.IsNotRegisteredError(err) {
-		klog.V(5).Infof("Unable to retrieve PartialObjectMetadata: %#v", err)
+		klog.FromContext(ctx).V(5).Info("Could not retrieve PartialObjectMetadata", "err", err)
```
Comment (mengjiao-liu) [inline staging/src/k8s.io/client-go/metadata/metadata.go:194]: The unit test(staging/src/k8s.io/client-go/metadata/metadata_test.go) corresponding to the `Get` method can use `_, ctx := ktesting.NewTestContext(t)` to pass context to test the contextual logging.

File: staging/src/k8s.io/client-go/metadata/metadata.go [hunk 2]
```diff
@@ -227,7 +227,7 @@ func (c *client) List(ctx context.Context, opts metav1.ListOptions) (*metav1.Par
 	}
 	obj, err := result.Get()
 	if runtime.IsNotRegisteredError(err) {
-		klog.V(5).Infof("Unable to retrieve PartialObjectMetadataList: %#v", err)
+		klog.FromContext(ctx).V(5).Info("Could not retrieve PartialObjectMetadataList", "err", err)
```
Comment (mengjiao-liu) [inline staging/src/k8s.io/client-go/metadata/metadata.go:230]: same above.

File: staging/src/k8s.io/client-go/metadata/metadata.go (hunk 1 above)
Comment (ricardoapl) [reply staging/src/k8s.io/client-go/metadata/metadata.go:194]: Thank you, should be resolved with 04e5522099623d66c1536ebc4253528280381c99

Please take another look

```
make test WHAT=./staging/src/k8s.io/client-go/metadata/ GOFLAGS="-v"
+++ [1208 10:21:40] Set GOMAXPROCS automatically to 12
+++ [1208 10:21:40] Running tests without code coverage and with -race
=== RUN   TestClient
=== RUN   TestClient/GET_is_able_to_convert_a_JSON_object_to_PartialObjectMetadata
    metadata.go:194: I1208 10:05:11.876937] Could not retrieve PartialObjectMetadata err="no kind \"Pod\" is registered for version \"v1\" in scheme \"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme/register.go:27\""
=== RUN   TestClient/LIST_is_able_to_convert_a_JSON_object_to_PartialObjectMetadata
    metadata.go:230: I1208 10:05:11.878102] Could not retrieve PartialObjectMetadataList err="no kind \"PodList\" is registered for version \"v1\" in scheme \"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme/register.go:27\""
=== RUN   TestClient/GET_fails_if_the_object_is_JSON_and_has_no_kind
=== RUN   TestClient/GET_fails_if_the_object_is_JSON_and_has_no_apiVersion
=== RUN   TestClient/GET_fails_if_the_object_is_JSON_and_not_clearly_metadata
    metadata.go:194: I1208 10:05:11.880156] Could not retrieve PartialObjectMetadata err="no kind \"Pod\" is registered for version \"v1\" in scheme \"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme/register.go:27\""
=== RUN   TestClient/Delete_fails_if_DeleteOptions_cannot_be_serialized_to_JSON
=== RUN   TestClient/DeleteCollection_fails_if_DeleteOptions_cannot_be_serialized_to_JSON
--- PASS: TestClient (0.01s)
    --- PASS: TestClient/GET_is_able_to_convert_a_JSON_object_to_PartialObjectMetadata (0.01s)
    --- PASS: TestClient/LIST_is_able_to_convert_a_JSON_object_to_PartialObjectMetadata (0.00s)
    --- PASS: TestClient/GET_fails_if_the_object_is_JSON_and_has_no_kind (0.00s)
    --- PASS: TestClient/GET_fails_if_the_object_is_JSON_and_has_no_apiVersion (0.00s)
    --- PASS: TestClient/GET_fails_if_the_object_is_JSON_and_not_clearly_metadata (0.00s)
    --- PASS: TestClient/Delete_fails_if_DeleteOptions_cannot_be_serialized_to_JSON (0.00s)
    --- PASS: TestClient/DeleteCollection_fails_if_DeleteOptions_cannot_be_serialized_to_JSON (0.00s)
PASS
ok      k8s.io/kubernetes/staging/src/k8s.io/client-go/metadata (cached)
```

File: staging/src/k8s.io/client-go/metadata/metadata.go (hunk 2 above)
Comment (ricardoapl) [reply staging/src/k8s.io/client-go/metadata/metadata.go:230]: Should also be resolved with 04e5522099623d66c1536ebc4253528280381c99

File: staging/src/k8s.io/client-go/metadata/metadata_test.go [hunk 3]
```diff
@@ -55,7 +56,7 @@ func TestClient(t *testing.T) {
 	testCases := []struct {
 		name    string
 		handler func(t *testing.T, w http.ResponseWriter, req *http.Request)
-		want    func(t *testing.T, client *Client)
+		want    func(t *testing.T, ctx context.Context, client *Client)
```
Comment (pohly) [inline staging/src/k8s.io/client-go/metadata/metadata_test.go:59]: Typically `ctx` is the first parameter in a function. I prefer to follow that also here and below.

**Comment by mengjiao-liu** [comment]
/lgtm
in my side

Wait for confirmation from the hack & client-go reviewer
/assign @sttts

File: staging/src/k8s.io/client-go/metadata/metadata_test.go (hunk 3 above)
Comment (pohly) [reply staging/src/k8s.io/client-go/metadata/metadata_test.go]: I agree, it's a bit subjective. I got annoyed by the separate parameters and proposed a solution here (still to be implemented, comments welcome): https://github.com/kubernetes/kubernetes/pull/122148#discussion_r1411874086

**Pull request merged**